pandas/core/indexes/datetimelike.py

```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
     _is_numeric_dtype = False
     _can_hold_strings = False
     _data: DatetimeArray | TimedeltaArray | PeriodArray
-    freq: BaseOffset | None
-    freqstr: str | None
-    _resolution_obj: Resolution
+
+    # ------------------------------------------------------------------------
+
+    @doc(DatetimeLikeArrayMixin.mean)
+    def mean(self, *, skipna: bool = True, axis: int | None = 0):
+        return self._data.mean(skipna=skipna, axis=axis)
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.asi8)
+    def asi8(self) -> npt.NDArray[np.int64]:
+        return self._data.asi8
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.freq)
+    def freq(self) -> BaseOffset | None:
+        return self._data.freq
+
+    @freq.setter
+    def freq(self, value) -> None:
+        # error: Property "freq" defined in "PeriodArray" is read-only
+        self._data.freq = value  # type: ignore[misc]
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.freqstr)
+    def freqstr(self) -> str | None:
+        return self._data.freqstr
+
+    # error: Decorated property not supported
+    @cache_readonly  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.inferred_freq)
+    def inferred_freq(self) -> str | None:
+        return self._data.inferred_freq
+
+    @cache_readonly
+    def _resolution_obj(self) -> Resolution | None:
```

Comment:
is None possible?

Reply:
`DatetimeLikeArrayMixin._resolution_obj` can theoretically be `None` but that is never the case when running the tests!

I could make sense to simplify `DatetimeLikeArrayMixin._resolution_obj` https://github.com/pandas-dev/pandas/blob/926b9ceff10d9b7a957811f0a4de3167332196de/pandas/core/arrays/datetimelike.py#L968 to

```py
    @property
    def _resolution_obj(self) -> Resolution:
        freqstr = self.freqstr
        return Resolution.get_reso_from_freqstr(freqstr)  # the KeyError is never triggered in the tests
```

Reply:
mh, in test pandas/tests/arrays/test_datetimes.py(82)test_fields(), `self.freq` is None but `_resolution_obj` is not called during this test but it would return `None` in this case. I came across this when testing whether `freqstr` can be `None` (and in this case it is `None`): if `freqstr` can be `None`, `_resolution_obj` can be `None`.

One argument for removing `None` from at least `_resolution_obj` (but not `freqstr`) is that places that call it, do not check for None and would anyways error there

Reply:
Yah, looks like it is only relevant for TDA, and we could re-use the logic (minus self.tz) from DTA._resolution_obj or that

Reply:
Sorry, I'm not following: do you suggest moving `_resolution_obj` (or `freqstr`) to a different class?

Reply:
`DatetimeLikeArrayMixin._resolution_obj` is only ever called for `PeriodArray` in the tests (other arrays might implement their own version?) and `DatetimeIndexOpsMixin._resolution_obj` is never called at least in the tests (I guess sub-classes implement it).

Cannot make `DatetimeIndexOpsMixin._resolution_obj` abstract as `DatetimeIndex` and `TimedeltaIndex` do not "implement" it (`DatetimeIndex` uses `inherit_names` to replace it but it didn't see anything for `TimedeltaIndex`).

Reply:
I would be fine leaving it as `Resolution | None` for now.

Reply:
if we don't need TDA._resolution_obj to exist, then we can just move it.  otherwise, i think something like

```
@property
def _resolution_obj(self) -> Resolution:
     tz = getattr(self, "tz", None)  # i.e. None for TimedeltaArray
    return get_resolution(self.asi8, tz, reso=self._reso)
```

works for DTA/TDA, and the current DatetimeLikeArrayMixin._resolution_obj becomes PeriodArray._resolution_obj

Reply:
Is `TimedeltaIndex` supposed to have `_resolution_obj` (and `resolution`):

```
>>> pd.TimedeltaIndex([1])._resolution_obj  # None
>>> pd.TimedeltaIndex([1]).resolution
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/properties.pyx", line 37, in pandas._libs.properties.CachedProperty.__get__
  File "python3.10/site-packages/pandas/core/indexes/extension.py", line 59, in cached
    return getattr(self._data, name)
  File "python3.10/site-packages/pandas/core/arrays/datetimelike.py", line 937, in resolution
    return self._resolution_obj.attrname  # type: ignore[union-attr]
AttributeError: 'NoneType' object has no attribute 'attrname'
```

If it is not supposed to have that, a good solution (for at least for the index classes) would be to make `DatetimeIndexOpsMixin._resolution_obj -> Resolution` abstract (all sub-classes except TimedeltaIndex implement it and return `Resolution`).